x-pack/plugins/observability_solution/apm/server/routes/assistant_functions/get_log_categories/index.ts

```diff
@@ -0,0 +1,157 @@
+ */
+
+import datemath from '@elastic/datemath';
+import { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { CoreRequestHandlerContext } from '@kbn/core/server';
+import { aiAssistantLogsIndexPattern } from '@kbn/observability-ai-assistant-plugin/server';
+import {
+  SERVICE_NAME,
+  CONTAINER_ID,
+  HOST_NAME,
+} from '../../../../common/es_fields/apm';
+import { getTypedSearch } from '../../diagnostics/create_typed_es_client';
+
+export type LogCategories =
+  | Array<{
+      key: string;
+      docCount: number;
+      sampleMessage: string;
+    }>
+  | undefined;
+
+export async function getLogCategories({
+  esClient,
+  coreContext,
+  arguments: args,
+}: {
+  esClient: ElasticsearchClient;
+  coreContext: CoreRequestHandlerContext;
+  arguments: {
+    start: string;
+    end: string;
+    'service.name'?: string;
+    'host.name'?: string;
+    'container.id'?: string;
+  };
+}): Promise<LogCategories> {
+  const start = datemath.parse(args.start)?.valueOf()!;
+  const end = datemath.parse(args.end)?.valueOf()!;
+
+  const keyValueFilters = getShouldMatchOrNotExistFilter([
+    { field: SERVICE_NAME, value: args[SERVICE_NAME] },
+    { field: CONTAINER_ID, value: args[CONTAINER_ID] },
+    { field: HOST_NAME, value: args[HOST_NAME] },
+  ]);
+
+  const index =
+    (await coreContext.uiSettings.client.get<string>(
+      aiAssistantLogsIndexPattern
+    )) ?? 'logs-*';
+
+  const search = getTypedSearch(esClient);
+
+  const query = {
+    bool: {
+      filter: [
+        ...keyValueFilters,
+        { exists: { field: 'message' } },
+        {
+          range: {
+            '@timestamp': {
+              gte: start,
+              lte: end,
+            },
+          },
+        },
+      ],
+    },
+  };
+
+  const hitCountRes = await search({
+    index,
+    size: 0,
+    track_total_hits: true,
+    query,
+  });
+  const totalDocCount = hitCountRes.hits.total.value;
+  const samplingProbability = Math.min(100_000 / totalDocCount, 1);
+
+  const categorizedLogsRes = await search({
+    index,
+    size: 0,
+    track_total_hits: 0,
+    query,
+    aggs: {
+      sampling: {
+        random_sampler: {
+          probability: samplingProbability,
+        },
+        aggs: {
+          categories: {
+            categorize_text: {
+              field: 'message',
+              size: 500,
+            },
+            aggs: {
+              sample: {
+                top_hits: {
+                  sort: [{ '@timestamp': 'desc' }],
+                  size: 1,
+                  _source: ['message'],
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  return categorizedLogsRes.aggregations?.sampling.categories?.buckets.map(
+    ({ doc_count: docCount, key, sample }) => {
+      // Supress error: Element implicitly has an 'any' type because expression of type '0' can't be used to index type 'SearchHitsMetadata<unknown>'.
+      // @ts-expect-error
```

Comment:
fwiw, this is probably because you don't have `'desc' as const`

Reply:
That doesn't make a difference. Could it be because I haven't defined a type for `_source`? I thought I could do something like `search<{ message: string }>({})` but that requires two arguments (the second one for the aggregations which I do not want to manually type)

Reply:
oh maybe because you haven't defined SubAggregateOf for the `categorize_text` aggregation type?

Reply:
never mind, looks like you did 😄

Reply:
let me have a look

Reply:
What you need is this:
```
sort: { '@timestamp': 'desc' as const },
```

and then indeed, you need to type `_source`, but you can do that in the map callback as well, `_source as x`.

Reply:
You are right, that did the trick!